### DISCUSSION TITLE
Possibility to add "comment" or "Note" on a specific day ?

### DISCUSSION THREAD

**Discussion by akhater** [iSoron/uhabits#194] [Feature Requests] [answered]
I will try to give an example, I have as habit to sleep at least 7 hours / night and last night I had an overnight at work so, obviously, I failed this habit, but I would like to have a way to add a note to that day explaining why I failed it would help when looking back to understand what happened on a particular day

thanks

**Comment by YPetremann** [comment]
Would be also be great for things like bloging or diversification habits

**Comment by foobarbazquxfred** [comment]
It would also be nice if any numeric "note" left could be graphed. So a user could see how many minutes he meditated this month, for example.

I would be willing to contribute to a bounty for the note feature, FWIW.

**Comment by ghost** [comment]
It will also be useful for swimmers or runners, in order to keep truck the time and distance.

**Comment by iSoron** [comment]
@Rightmanj This will be covered by numerical habits (see #42)

**Comment by holocircuit** [comment]
I'd also like this, and would be willing to write it.

I'm not sure about the best way to do it with the interface though. Perhaps "super-long press" should bring up a menu with an option to add a note?

**Comment by tickTackHack** [comment]
I'm also willing to give it a shot to implement this.  I would recommend making the input length restricted to a relatively short length. Maybe 240 characters? This can help with keeping the database size small and help deter security issues.

**Comment by c33s** [comment]
not sure it makes sense to restrict the length of the field. i personally see no benefit on having a "small"  database. it looks like the performance impact is not that huge (should be tested on an android device before of course). the devices all have quite a lot of space, so keeping the size small does not really matter.  
even if someone writes a long text to 10 habbits, it is still only text. so it is not really big in size.
 
> I recognized, that performance drastically increases with table size dropping. I found:
> 
>     1.6M entries 327 ms w/o indexing => 29.7 ms with indexing
>     670k entries 149 ms w/o indexing => 28.8 ms with indexing
>     280k entries 71 ms w/o indexing => 28.5 ms with indexing
>     147k entries 44 ms w/o indexing => 28.0 ms with indexing
>     19k entries 25 ms w/o indexing => 25.0 ms with indexing
https://stackoverflow.com/q/33456278

> I've created SQLite databases up to 3.5GB in size with no noticeable performance issues. If I remember correctly, I think SQLite2 might have had some lower limits, but I don't think SQLite3 has any such issues.
> 
> According to the SQLite Limits page, the maximum size of each database page is 32K. And the maximum pages in a database is 1024^3. So by my math that comes out to 32 terabytes as the maximum size. I think you'll hit your file system's limits before hitting SQLite's!
https://stackoverflow.com/a/786146

some sort of export text or archive text (to delete the text for a certain period) would be nice.

**Comment by fiosmiles** [comment]
Would love this, just started using the app and the ability to comment on a daily habit (along with the idea of "not done" which I see under discussion),  e.g. flossing habit - "travelling and forgot to bring floss!" And being able to access the notes in a journal style when you open a habit. The calendar could have a small icon modified to indicate there is a note.

**Comment by kpuljek** [comment]
+1 for this, i'm trying to watch more movies and it would be great if I could add their name to individual entries

**Comment by doublehelix24** [comment]
Another +1. Would love this and used this feature a lot in HabitBull.

**Comment by nottheend** [comment]
+1
It would be beneficial in cases, where a habbit couldn't be placed on a specific day. It is helpful on retrospectives to understand better the (missing) progress of the habbit

**Comment by gaborod16** [comment]
+1
Really love the App, I would even contribute to this feature if it gets into the planning.

**Comment by KristianTashkov** [comment]
@iSoron If we ever merge the advanced options dialog, notes become a very easy continuation of it. I originally planned to implement it after we merge the dialog into master.

**Comment by Mrnofish** [comment]
Please upvote this suggestion to signal your interest: the control is next to the discussion's title up above.

**Comment by iSoron** [comment, marked as answer]
This has been implemented by @vbh (see #1103) and should become available in Loop 2.1.

**Reply by Mrnofish** [reply]
Very cool, thank you @vbh and @iSoron

**Reply by nottheend** [reply]
Is there a way to "migrate" existing habits to a numeric way?
Thank you so much, this is really great now!

**Reply by driabchenko** [reply]
Is there a way to enable this on Android?

**Reply by Mrnofish** [reply]
@driabchenko long tap on a day from the main screen, you'll get a selection dialog that also includes a field for adding notes.

**Reply by driabchenko** [reply]
Thanks! I found it. In case someone else will come looking, I had to become a "beta tester" of the app in Play Store on Android to be able to update to version 2.1 and use this feature. As of today, the default version available from Play Store is 2.0 which does not support notes.

**Reply by ConorSheehan1** [reply]
@iSoron ~would it be possible to add per day notes to boolean habits too? Is there a feature request for that already or should I create one?~

My bad, just accidentally found how to add notes to boolean habits by double tapping.

**Reply by ConorSheehan1** [reply]
@nottheend  I don't think there's an official way but I made a little python script to convert from boolean to numeric https://github.com/ConorSheehan1/uhabits_converter

**Comment by vnagel** [comment]
Is there a plan to include these notes in the CSV export?
